application/ledger_query_service/src/ethereum/queries/event_query.rs

```diff
@@ -0,0 +1,507 @@
+}
+
+/// Event Matcher work similar as web3 filters:
+/// https://web3js.readthedocs.io/en/1.0/web3-eth-subscribe.html?highlight=filter#subscribe-logs
+/// E.g. this `EventMatcher` would match this `Log`:
+/// ```
+/// EventMatcher {
+/// address: 0xe46FB33e4DB653De84cB0E0E8b810A6c4cD39d59,
+/// data: None,
+/// topics: [
+/// None,
+/// Some(0x000000000000000000000000e46fb33e4db653de84cb0e0e8b810a6c4cd39d59),
+/// None()
+/// ],
+/// ```
+/// ```
+/// Log:
+/// [ { address: '0xe46FB33e4DB653De84cB0E0E8b810A6c4cD39d59',
+/// data: '0x123',
+/// ..
+/// topics:
+/// [ '0xddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3ef',
+/// '0x000000000000000000000000e46fb33e4db653de84cb0e0e8b810a6c4cd39d59',
+/// '0x000000000000000000000000d51ecee7414c4445534f74208538683702cbb3e4' ],
+/// },
+/// .. ] //Other data omitted
+/// }
+/// ```
+#[derive(Debug, Clone, Serialize, Deserialize, Default)]
+struct EventMatcher {
+    address: Option<Address>,
+    data: Option<Bytes>,
+    topics: Vec<Option<Topic>>,
+}
+
+impl EventQuery {
+    pub fn matches_block(&self, block: &Block<Transaction>) -> bool {
+        match self {
+            Self { event_matchers } => event_matchers.iter().all(|log_matcher| {
+                log_matcher.topics.iter().all(|topic| {
+                    topic.as_ref().map_or(true, |topic| {
+                        block.logs_bloom.contains_input(Input::Raw(&topic.0))
+                    })
+                })
+            }),
+        }
+    }
+
+    pub fn matches_transaction_receipt(&self, transaction_receipt: TransactionReceipt) -> bool {
+        match self {
+            Self { event_matchers } => {
+                event_matchers
+                    .iter()
+                    .all(|event_matcher| match event_matcher {
+                        EventMatcher {
+                            address: None,
+                            data: None,
+                            topics,
+                        } if topics.is_empty() => false,
+                        EventMatcher {
+                            address,
+                            data,
+                            topics,
+                        } => {
+                            transaction_receipt.logs.iter().any(|tx_log| {
+                                let mut result = true;
+
+                                if address
+                                    .as_ref()
+                                    .map_or(false, |address| address != &tx_log.address)
+                                {
+                                    return false;
+                                }
+
+                                if data.as_ref().map_or(false, |data| data != &tx_log.data) {
+                                    return false;
+                                }
+
+                                if tx_log.topics.len() == topics.len() {
+                                    tx_log.topics.iter().enumerate().for_each(
```

Comment:
🔧 Just make this `foreach` an `all` and I think you can completely delete this `result` variable.

Reply:
true, good point 👍